config/config.go

```
@@ -263,6 +262,167 @@ func FromViper(v *viper.Viper) (*Config, error) {
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return "", fmt.Errorf("failed to create stdout pipe for tree-root-cmd: %w", err)
+	}
+
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return "", fmt.Errorf("failed to create stderr pipe for tree-root-cmd: %w", err)
+	}
+
+	// start processing stderr before we begin executing the command
+	go func() {
+		// capture stderr line by line and log
+		l := log.WithPrefix("tree-root-cmd | stderr")
+
+		scanner := bufio.NewScanner(stderr)
+		for scanner.Scan() {
+			l.Debugf("%s", scanner.Text())
+		}
+	}()
+
+	// start executing without waiting
+	if cmdErr := cmd.Start(); cmdErr != nil {
+		return "", fmt.Errorf("failed to start tree-root-cmd: %w", cmdErr)
+	}
+
+	// read stdout until it is closed (command exits)
+	output, err := io.ReadAll(stdout)
+	if err != nil {
+		return "", fmt.Errorf("failed to read stdout from tree-root-cmd: %w", err)
+	}
+
+	log.WithPrefix("tree-root-cmd | stdout").Debugf("%s", output)
+
+	// check execution error
+	if cmdErr := cmd.Wait(); cmdErr != nil {
+		var exitErr *exec.ExitError
+
+		// by experimenting, I noticed that sometimes we received the deadline exceeded error first, other times
+		// the exit error indicating the process was killed, therefore, we look for both
+		tookTooLong := errors.Is(cmdErr, context.DeadlineExceeded)
+		tookTooLong = tookTooLong || (errors.As(cmdErr, &exitErr) && exitErr.ProcessState.String() == "signal: killed")
+
+		if tookTooLong {
+			return "", fmt.Errorf(
+				"tree-root-cmd was killed after taking more than %v to execute",
+				executionTimeout,
+			)
+		}
+
+		// otherwise, some other kind of error occurred
+		return "", fmt.Errorf("failed to execute tree-root-cmd: %w", cmdErr)
+	}
+
+	// trim the output and check it's not empty
+	treeRoot := strings.TrimSpace(string(output))
+
+	if treeRoot == "" {
+		return "", fmt.Errorf("empty output received after executing tree-root-cmd: %s", cfg.TreeRootCmd)
+	}
```

I personally would prefer to put this check up a layer in `determineTreeRoot`. If somebody explicitly specifies `--tree-root ""`, I'd also like them to get this error, rather than having it fall back to the current directory.

---

> If somebody explicitly specifies `--tree-root " "`

I was about to comment that `""` would be treated as `tree-root` being undefined, however if a whitespace-only value is defined then you're right, it _would_ run into this edge-case.

---

Oops, good point. I didn't remember we're treating "unset" and "set to empty string" as the same thing here.

I still do feel like it would make more sense to do this check as part of the "is this potential root a valid path? let's normalize it" codepath, and not in the implementation of a specific root discovery mechanism.

---

I'm not quite following what is being advocated for here.

---

> IMO this begs the question of whether trimming stdout is the correct approach anyway.

In hindsight, I believe the most correct approach here would be to split on `\n`, then remove the empty lines.

You can then assert that the remaining line count is 1, and also use the lines slice to log the stdout line-by-line.

```go
lines := strings.Split(stdout, "\n")
nonEmptyLines := slices.DeleteFunc(lines, func(line string) bool {
	return line == ""
}

switch (len(nonEmptyLines)) {
case 1:
	// no-op

case 0:
	return "", fmt.Errorf("empty output received after executing tree-root-cmd: %s", cfg.TreeRootCmd)

default:
	// Log all stdout line-by-line, including the blank lines
	logger := log.WithPrefix("tree-root-cmd | stdout")
	for _, line := range lines {
		logger.Error(line)
	}
	return "", fmt.Errorf("tree-root-cmd cannot output multiple lines: %s", cfg.TreeRootCmd)
}
```

---

As said before, trimming is fine for the initial impl, I can open a follow up PR for this if preferred.

---

All mechanisms must pass through the same stage of realising their output as absolute paths, which should throw an error for any malformed paths. Trimming the output of the command execution and enforcing a single line seems prudent, specifically for the `tree-root-cmd` pathway.

---

> This is the part that bothers me: IIRC, we're using go's built-in `Path.abs` to do this, which accepts empty strings rather than erroring.

I believe it is impossible to define a _truly_ empty tree-root path; attempting to set `tree-root = ""` will be treated as undefined. Returning an empty string from `tree-root-cmd` will be caught here.

As for blank/whitespace paths, a whitepace string is technically a valid relative path. On some filesystems you can name a file/dir `" "` (if you really wanted to).

This is one reason why the current impl splits stdout into lines, instead of just trimming it.